identity-wallet/src/state/credentials/reducers/send_token_request.rs

```diff
@@ -79,9 +81,85 @@ pub async fn send_token_request(state: AppState, action: Action) -> Result<AppSt
+                        "The specified authorization server {specified_auth_server} is not listed in the credential issuer metadata."
+                    )));
+                        }
+                    }
+                }
+            }
+        } else if let Some(grants) = &credential_offer.grants {
+            if let Some(auth_code) = &grants.authorization_code {
+                if let Some(specified_auth_server) = &auth_code.authorization_server {
+                    if !credential_issuer_metadata.authorization_servers.is_empty()
+                        && !credential_issuer_metadata
+                            .authorization_servers
+                            .contains(specified_auth_server)
+                    {
+                        return Err(AppError::Error(format!(
+                                "The specified authorization server {specified_auth_server} is not listed in the credential issuer metadata."
+                            )));
+                    }
+                }
+            }
+        }
+
+        // Extract the authorization server selection from the authorization_server parameter in the grant types.
+        let authorization_server_url = if is_pre_authorized {
+            credential_offer
+                .grants
+                .as_ref()
+                .and_then(|grants| grants.pre_authorized_code.as_ref())
+                .and_then(|pre_authorized_code| pre_authorized_code.authorization_server.as_ref())
+                .cloned()
+                .or_else(|| {
+                    // If no authorization server is specified, fall back to the authorization_servers in the credential issuer metadata.
+                    if !credential_issuer_metadata.authorization_servers.is_empty() {
+                        // Select the first authorization server in the array.
+                        Some(credential_issuer_metadata.authorization_servers[0].clone())
+                    } else {
+                        None
+                    }
+                })
+                // Fall back to credential issuer url if no authorization server is specified.
+                .unwrap_or_else(|| credential_issuer_url.clone())
+        } else {
+            credential_offer
+                .grants
+                .as_ref()
+                .and_then(|grants| grants.authorization_code.as_ref())
+                .and_then(|auth_code| auth_code.authorization_server.as_ref())
+                .cloned()
+                .or_else(|| {
+                    if !credential_issuer_metadata.authorization_servers.is_empty() {
+                        Some(credential_issuer_metadata.authorization_servers[0].clone())
+                    } else {
+                        None
+                    }
+                })
+                // Fall back to credential issuer url if no authorization server is specified.
+                .unwrap_or_else(|| credential_issuer_url.clone())
+        };
```

Comment:
The `if` and `else` blocks are nearly identical; only the grants differ, so see if you can 'move' the `if is_pre_authorized { .. } else { .. }` logic inside an `and_then` to keep the duplicated code at a minimum. :)